File: AoC2017/Day01.fs
namespace AoC2017

open Common
open System.Collections.Generic
open System.Linq

type Day01 = 
    inherit Solution

    new (file:string) = { inherit Solution(file) }
    new (input : List<string>) = { inherit Solution(input) } 

    override this.SolvePart1() = 
        let input = this.Input.First()
        let length = input.Length - 1
        let mutable sum = 0.0
        for i in 0..length do
            let next = if i + 1 <= length then input.[i + 1] else input.[0]
            if input.[i] = next then sum <-  sum + System.Char.GetNumericValue input.[i]
        
        sum.ToString()

    override this.SolvePart2() = 
        let input = this.Input.First()
        let length = input.Length 
        let mutable sum = 0.0
        for i in 0..length-1 do
            let next = input.[(i + length/2) % length]
            if input.[i] = next then sum <-  sum + System.Char.GetNumericValue input.[i]
        
        sum.ToString()
 

File: AoC2017/Day01.cs
using Common;
using Common.Extensions;

namespace AoC2017;

public class Day01 : Solution
{
	private CircularList<int> digits;
	public Day01(string file) : base(file) => InputToDigits( );

	public Day01(List<string> input) : base(input) => InputToDigits( );

	private void InputToDigits()
	{
		digits = Input.First( ).Aggregate(new CircularList<int>( ), (list, c) =>
		{
			list.Add(c.ToInt( ));
			return list;
		});
		digits.ResetHead( );
	}

	public override async Task<string> SolvePart1() => digits
		.Aggregate(0, (sum, i) =>
		{
			digits.MoveRight( );
			return i == digits.Current ? sum + i : sum;
		}).ToString( );


	public override async Task<string> SolvePart2() => digits.WithIndex()
		.Aggregate(0, (sum, i) =>
		{
			digits.SetHeadByIndex(i.idx);
			digits.MoveRight(digits.Count / 2);
			return i.Value == digits.Current ? sum + i.Value : sum;
		}).ToString( );
}
